device/Microsoft.Azure.Devices.Client/DeviceAuthenticationWithTpm.cs

```diff
@@ -0,0 +1,105 @@
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Microsoft.Azure.Devices.Shared;
+using System;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.Devices.Client
+{
+    /// <summary>
+    /// Authentication method that uses a shared access signature token and allows for token refresh. 
+    /// </summary>
+    public sealed class DeviceAuthenticationWithTpm : DeviceAuthenticationWithTokenRefresh
+    {
+        private SecurityClientHsmTpm _securityClient;
+        private int _timeToLiveSeconds;
+
+        public DeviceAuthenticationWithTpm(
+            string deviceId, 
+            SecurityClientHsmTpm securityClient,
+            int timeToLiveSeconds = 1 * 60 * 60,
+            int timeToLiveBufferSeconds = 10 * 60) : base(deviceId, timeToLiveBufferSeconds)
+        {
+            if (securityClient == null)
+            {
+                throw new ArgumentNullException(nameof(securityClient));
+            }
+
+            if (timeToLiveSeconds < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(timeToLiveSeconds));
+            }
+
+            if (timeToLiveBufferSeconds < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(timeToLiveBufferSeconds));
+            }
+
+            _securityClient = securityClient;
+            _timeToLiveSeconds = timeToLiveSeconds;
+        }
+        
+        protected override Task<string> SafeCreateNewToken(string iotHubName)
+        {
+            const long WINDOWS_TICKS_PER_SEC = 10000000;
+            const long EPOCH_DIFFERNECE = 11644473600;
+            long expirationTime = (DateTime.Now.ToUniversalTime().ToFileTime() / WINDOWS_TICKS_PER_SEC) 
+                                  - EPOCH_DIFFERNECE + _timeToLiveSeconds;
+
+            string sasToken = "";
+            if ((iotHubName.Length > 0) && (DeviceId.Length > 0))
+            {
+                // Encode the message to sign with the TPM
+                UTF8Encoding utf8 = new UTF8Encoding();
+                string tokenContent = iotHubName + "/devices/" + DeviceId + "\n" + expirationTime;
+                byte[] encodedBytes = utf8.GetBytes(tokenContent);
+
+                // Sign the message
+                byte[] hmac = _securityClient.Sign(encodedBytes);
+
+                // If we got a signature format it
+                if (hmac.Length > 0)
+                {
+                    // Encode the output and assemble the connection string
+                    string hmacString = AzureUrlEncode(System.Convert.ToBase64String(hmac));
+                    sasToken = "SharedAccessSignature sr=" + iotHubName + "/devices/" + DeviceId + "&sig=" + hmacString + "&se=" + expirationTime;
+                }
+            }
+
+            return Task.FromResult<string>(sasToken);
+        }
+
+        private string AzureUrlEncode(string stringIn)
```

Comment:
>AzureUrlEncode [](start = 23, length = 14)

Any reason why building your own instead of using one of the ones from the .Net lib?

Reply:
Good point. @simonporter - is there anything special in the `AzureUrlEncode` different from the URI RFCs?